### PR TITLE
Custom GPU driver UI (Adrenotools UI)

### PR DESCRIPTION
**Pull request by lynxnb**
It was already possible to use custom GPU drivers in Skyline, although it required recompiling the app every time, limiting this feature to advanced users only. This PR adds a preference and an activity in the app, allowing for custom GPU drivers management. The functionality is only available on devices with a Qualcomm Adreno GPU. For all other devices, the setting will be greyed out and disabled.

TODO before merging:
- [x] Merge #324
- [x] Rebase

### CODE REVIEW

**Comment by Joko1994** [comment]
Fps more drop on game dragon star varnir. It's only 2-3 fps. The old version of I have give 15 fps. And still have many graphics
)) glichy and blur @lynxnb [](url)

**Comment by mateoalfaro** [comment]
> Fps more drop on game dragon star varnir. It's only 2-3 fps. The old version of I have give 15 fps. And still have many graphics )) glichy and blur @lynxnb

If different drivers cause issues for that game you could just use the stock driver when playing that game

**Comment by lynxnb** [comment]
@Joko1994 don't report issues related to custom drivers' performance here. Performance differences given by their use is expected and nothing we can do about it.

**Comment by AbandonedCart** [comment]
@lynxnb Is there documentation somewhere on the structure of the archives compatible with this?

**Comment by lynxnb** [comment]
Not yet that I'm aware of, we'll make it available when it's time.

**Comment by AbandonedCart** [comment]
Fair enough. As long as they are available internally, I guess it isn't really crucial until a merge.

**Comment by bylaws** [comment]
> Fair enough. As long as they are available internally, I guess it isn't really crucial until a merge.

See adrenotools GitHub, there are docs there

**Comment by AbandonedCart** [comment]
> See adrenotools GitHub, there are docs there

Maybe I worded the question poorly. What I was asking if this is meant to load the packages currently being installed through Magisk or is this going to be using an alternate format.

**Comment by mr-hopeless** [comment]
> Maybe I worded the question poorly. What I was asking if this is meant to load the packages currently being installed through Magisk or is this going to be using an alternate format.
> 
> 

Alternate format

**Comment by PixelyIon** [comment]
@AbandonedCart The format aims to be compatible with Magisk and will have the extension `filename.adpkg.zip` in the future, it should both be able to be directly installed by Magisk or used by the UI in Skyline.

**Comment by AbandonedCart** [comment]
I was hoping that was the eventual intention, but figured it wasn't worth creating any confusion with it clearly stated that this particular version is not compatible.

**Comment by sakazuk** [comment]
adreno 650 pls

**Comment by mr-hopeless** [comment]
> adreno 650 pls
> 

It's available for it, my dear

File: app/src/main/cpp/skyline/gpu.cpp [hunk 1]
```diff
@@ -332,30 +333,27 @@ namespace skyline::gpu {
...
-            nullptr
-        )};
-        if (!libvulkanHandle) {
+        void *libvulkanHandle{};
+
+        // If the user has selected a custom driver, try to load it
+        if (!(*state.settings->gpuDriver).empty()) {
             libvulkanHandle = adrenotools_open_libvulkan(
                 RTLD_NOW,
-                ADRENOTOOLS_DRIVER_FILE_REDIRECT,
```
Comment (bylaws) [inline app/src/main/cpp/skyline/gpu.cpp:348]: you should do  ADRENOTOOLS_DRIVER_FILE_REDIRECT | ADRENOTOOLS_DRIVER_CUSTOM and file redirect into the public data dir

File: app/src/main/cpp/skyline/gpu.cpp [hunk 2]
```diff
@@ -332,30 +333,27 @@ namespace skyline::gpu {
...
-                nullptr,
-                (state.os->publicAppFilesPath + "gpu/vk_file_redirect/").c_str()
+                (state.os->privateAppFilesPath + "gpu_drivers/" + *state.settings->gpuDriver + "/").c_str(),
+                (*state.settings->gpuDriverLibraryName).c_str(),
+                nullptr
             );
+
             if (!libvulkanHandle)
-                libvulkanHandle = dlopen("libvulkan.so", RTLD_NOW);
+                Logger::Warn("Failed to load custom Vulkan driver: {}/{}", *state.settings->gpuDriver, *state.settings->gpuDriverLibraryName);
```
Comment (bylaws) [inline app/src/main/cpp/skyline/gpu.cpp:351]: print dlerror here

**Review by bylaws — CHANGES_REQUESTED** [review]

File: app/src/main/cpp/driver_jni.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#include <vulkan/vulkan_raii.hpp>
+#include "skyline/common/signal.h"
+#include "skyline/common/utils.h"
+
+extern "C" JNIEXPORT jobjectArray JNICALL Java_emu_skyline_utils_GpuDriverHelper_00024Companion_getSystemDriverInfo(JNIEnv *env, jobject) {
+    auto libvulkanHandle{dlopen("libvulkan.so", RTLD_NOW)};
+
+    vk::raii::Context vkContext{reinterpret_cast<PFN_vkGetInstanceProcAddr>(dlsym(libvulkanHandle, "vkGetInstanceProcAddr"))};
+    vk::raii::Instance vkInstance{vkContext, vk::InstanceCreateInfo{}};
+    vk::raii::PhysicalDevice physicalDevice{std::move(vk::raii::PhysicalDevices(vkInstance).front())}; // Get the first device as we aren't expecting multiple GPUs
```
[the remark this answers is not in the document]
Comment (lynxnb) [reply app/src/main/cpp/driver_jni.cpp:16]: It won't compile without

File: app/src/main/cpp/skyline/gpu.cpp (hunk 1 above)
Comment (lynxnb) [reply app/src/main/cpp/skyline/gpu.cpp:348]: > file redirect into the public data dir
Could you elaborate a bit?

Comment (bylaws) [reply app/src/main/cpp/skyline/gpu.cpp:348]: uh just do what was there before, pass in dir in the last param and pass in flags as `ADRENOTOOLS_DRIVER_FILE_REDIRECT | ADRENOTOOLS_DRIVER_CUSTOM`

File: app/src/main/cpp/skyline/gpu.cpp [hunk 4]
```diff
@@ -332,30 +333,27 @@ namespace skyline::gpu {
...
-                nullptr,
-                nullptr,
+                (state.os->privateAppFilesPath + "gpu_drivers/" + *state.settings->gpuDriver + "/").c_str(),
+                (*state.settings->gpuDriverLibraryName).c_str(),
                 (state.os->publicAppFilesPath + "gpu/vk_file_redirect/").c_str()
             );
+
             if (!libvulkanHandle)
-                libvulkanHandle = dlopen("libvulkan.so", RTLD_NOW);
+                Logger::Warn("Failed to load custom Vulkan driver {}/{}: {}", *state.settings->gpuDriver, *state.settings->gpuDriverLibraryName, dlerror());
```
Comment (bylaws) [inline app/src/main/cpp/skyline/gpu.cpp:351]: dlerror can return null, this might be fine but worth checking, otherwise lgtm

Comment (lynxnb) [reply app/src/main/cpp/skyline/gpu.cpp:351]: Oh that's why it was crashing, thanks

File: app/src/main/cpp/driver_jni.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    auto deviceProperties2{physicalDevice.getProperties2<vk::PhysicalDeviceProperties2, vk::PhysicalDeviceDriverProperties>()};
+    auto properties{deviceProperties2.get<vk::PhysicalDeviceProperties2>().properties};
+
+    auto driverId{vk::to_string(deviceProperties2.get<vk::PhysicalDeviceDriverProperties>().driverID)};
+    auto driverVersion{skyline::util::Format("{}.{}.{}", VK_API_VERSION_MAJOR(properties.driverVersion), VK_API_VERSION_MINOR(properties.driverVersion), VK_API_VERSION_PATCH(properties.driverVersion))};
+
+    auto array = env->NewObjectArray(2, env->FindClass("java/lang/String"), nullptr);
```
Comment (bylaws) [inline app/src/main/cpp/driver_jni.cpp:24]: Use {}

**Review by PixelyIon — APPROVED** [review]
This PR merge is being fast-tracked, there are minor style inconsistencies that are being glossed over and will be fixed post-merge.

**Review by bylaws — APPROVED** [review]
FastTrack merge approved

**Pull request merged**